js/packages/binderhub-client/lib/index.js

```diff
@@ -118,30 +118,30 @@ export class BinderRepository {
     // Make a copy of the URL so we don't mangle the original
     let url = new URL(serverUrl);
     if (path) {
-      // strip trailing / from URL
-      url.pathname = url.pathname.replace(/\/$/, "");
-
+      // Ensure there is a trailing / in serverUrl
+      if (!url.pathname.endsWith("/")) {
+        url.pathname += "/";
+      }
```

Comment:
I think this is redudant when using the URL objects, but I'm not 100%.

```javascript
let baseUrl = new URL("https://developer.mozilla.org");
console.log(new URL("doc/tree/?test=ok", baseUrl))
```

![image](https://github.com/jupyterhub/binderhub/assets/3837114/7e132d63-f31a-41dc-9a7b-5d80d40037e4)

Do we delete it or retain it?

```suggestion
```

Reply:
Hmm okay so pathname may sometimes not endWith `/`, for example...

![image](https://github.com/jupyterhub/binderhub/assets/3837114/0fcb1c62-ff2f-47fe-970e-6ed335351ea2)

But it seems that the path is ignored when using this URL to construct a new URL via the URL constructor that includes a baseUrl argument.

![image](https://github.com/jupyterhub/binderhub/assets/3837114/33f13c6f-aff3-4c00-8ad1-206d348d376c)

Reply:
Ah!

I think there is a breaking change here.

```diff
-url.pathname = url.pathname + "/doc/tree/" + encodeURI(path);
+url = new URL("doc/tree/" + encodeURI(path), url);
```

These are different I think, to verify that:

```
console.log("old new behavior presented below")

let nonRoot1 = new URL("https://developer.mozilla.org/test")
nonRoot1.pathname = nonRoot1.pathname.replace(/\/$/, "");
nonRoot1.pathname = nonRoot1.pathname + "/doc/tree/"
console.log(nonRoot1.href)

let root1 = new URL("https://developer.mozilla.org")
root1.pathname = root1.pathname.replace(/\/$/, "");
root1.pathname = root1.pathname + "/doc/tree/"
console.log(root1.href)

console.log("new behavior comes below")

let nonRoot2 = new URL("https://developer.mozilla.org/test")
let newNonRoot2 = new URL("doc/tree", nonRoot2)
console.log(newNonRoot2.href)

let root2 = new URL("https://developer.mozilla.org")
let newRoot2 = new URL("doc/tree", root2)
console.log(newRoot2.href)
```

![image](https://github.com/jupyterhub/binderhub/assets/3837114/cd5e74aa-35c6-45a0-bb69-16a6ce22b77b)

Reply:
The trailing `/` in the pathname is required (`/test/` not `/test`), it affects how the parts are joined.

Reply:
Yeah, preventing this behavior is why we enforce path ends with `/` even if it is not passed in. So baseUrl will never not have trailing slash. As @manics says, it affects how parts are joined.

Reply:
@consideRatio I added another test case to demonstrate why I think this is fine.

Reply:
Hmmm but what if baseUrl is a jupyterhub with a prefix?

Note that now, but not before, the nonRoot examples doesnt have the /test prefix in the path?

Reply:
If you both think there is no loss of path issue, this is good even if i dont understand the details.

Reply:
Ah, so the href can be the same for a URL object, but end up resulting in a different URL when used to construct another. I think i got it and this LGTM